UI Page Layout: Layout and MainControls, Main- and Metabar, Slates

nhaagen opened the pull request:
Dear All,

this is the current state of work on https://docu.ilias.de/goto_docu_wiki_wpage_4563_1357.html
(respectively https://docu.ilias.de/goto_docu_wiki_wpage_5095_1357.html) with the
goal to realize the overall page layout as sketched here: https://sketch.cloud/s/yLndE

This PR introduces:
 - [Two new Glyphs ("Search" and "Help")](https://github.com/ILIAS-eLearning/ILIAS/pull/1335/files#diff-636dc2f66a0722b4e790c5654824ddc1)
 - The [Maincontrols](https://github.com/ILIAS-eLearning/ILIAS/pull/1335/files#diff-19b7a6034a22226f59300948731bc31f) "Mainbar" and "Metabar" with Slates ([Prompts](https://github.com/ILIAS-eLearning/ILIAS/pull/1335/files#diff-19a691ed997016a8b06a2ea3e368819d))
 - The [Standard Page](https://github.com/ILIAS-eLearning/ILIAS/pull/1335/files#diff-ddc95be6d038d5279fc6125f12f14745) as Layout Component.

Thanks for your considerations and comments.

nhaagen committed:
UI Page Layout: Layout and MainControls, Main- and Metabar, Slates

alex40724 commented on src/UI/Component/MainControls/Slate/Slate.php:
```diff
@@ -0,0 +1,50 @@
+	public function getToggleSignal(): Signal;
+
+	/**
+	 * Signal that enagages the slate when tiriggered.
```
Some typos here. And I think this needs some detailed explanation (if its not somewhere else in the doc). This could be used to load the slate content "on demand" per ajax, right? I think this will be necessary already in the first version to work well, since we may have several tools/slates that will do extensive querying for their presentation (e.g. who is online). It is important that this will fire also if the slate is initially presented (e.g. a tool on top) withouth any button being pressed. And also if this fires if a slate is toggled multiple times or if this only fires on the first "show" event. I guess its the first, so consumers must take care not to render twice.

nhaagen replied:
In fact, the Show-Signal is very close to the Toggle-Signal, only, that it just opens the Slate and does never close it (this is for the Slates in the Mainbar's tool-section). We could probably move this to the JS-Layer as well...

atoedt replied:
Oliver and I would need the drawn output of this factory. Could you please provide drawn output / link to a Kitchesink that is publicly accessible?  Thank you

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *   purpose: >
+	 *     The Metabar is a unique page section to accomodate elements that
+	 *     should permamently be in sight of the user.
+	 *     The Metabar shall, first of all, host Prompts, i.e. notifications
```
It would be nice to have this represented in the examples. Currently the example uses bulky buttons only.

Amstutz replied:
I agree that there should be one example, that is as closely to the look some standard ILIAS will cary. Especially in this state of the discussion, this could help. Nevertheless, I also like to somewhat more abstract example, since they help to really focus on the general concept.

alex40724 commented on src/UI/Component/Layout/Page/Factory.php:
```diff
@@ -0,0 +1,46 @@
+	 *    The Standard Page is the regular view upon ILIAS.
+	 *
+	 *   composition: >
+	 *      The main parts of a Page are the Metabar hosting the logo and tools,
```
I think it would be worth discussing, if the logo should be part of the metabar or not. The alternative would be to pass the logo directly to the Page Layout. This would allow to locate the logo at a place more distant from the metabar (of course the page layout could "extract" the logo from the metabar but this feels somehow wrong).

klees replied:
Sounds plausible to me, what do you think @nhaagen?

nhaagen replied:
All fine with me.

alex40724 commented on src/UI/examples/Layout/Page/Standard/ui.php:
```diff
@@ -0,0 +1,44 @@
+	);
+	$breadcrumbs = $f->breadcrumbs($crumbs);
+
+	$content = pagedemoContent($f);
```
Personally I would prefer to have this functions in the same example file and not to use includes. I think this would increase readability. But this may be a matter of taste.

alex40724 commented on src/UI/templates/js/MainControls/mainbar.js:
```diff
@@ -0,0 +1,213 @@
+			,_cls_btn_engaged = 'engaged'
+			,_cls_page_active_slates = 'with-mainbar-slates-engaged' //set on _page_div
+			,_cls_entries_wrapper = 'il-mainbar-entries' //encapsulating div of all entries
+			,_cls_toolentries_wrapper = 'il-mainbar-tools-entries' //tools (within  _cls_entries_wrapper)
```
I wonder if we should have some naming rules that distinguish style classes that are referenced in JS from style classes that are for design purposes only. Somthing that tells the person that modifies a template: Changing this name will break the behaviour.

klees replied:
Sounds very plausible and would also be applicable in other contexts (looking at you, filter), right? This would imo also require that a class is either for styling or for behaviour, not both. Should we take this discussion elsewhere, maybe into a proposal for the roadmap of the UI-framework? Does someone feel entitle to write a proposal?

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *
+	 *     Slates are not part of the content and will reside next to or over it. They
+	 *     will open and close without changing the current context.
+	 *     Accordingly, Slates depend on a component that toggles their visibility.
```
I find it hard to understand if the triggerer and its properties is an external or internal thing to the slate. Who is taking care of icon/caption? Usually (like with buttons and popovers) the content container (popover) is separated from the triggerer. The current "Legacy Slate" example illustrates this unclear responsiblity imo. A triggerer (bulky button) has an icon/symbol and name/caption. But this is obtained from the slate object. What if they use different icons?

klees replied:
We model it like this: The label and the icon(or glyph) belong to the slate, while the component that uses slates needs to define how these two are wrapped into a concrete triggerer (i.e. the button). We think, that the icon/glyph and label need to belong to the slate, because they are deeply connected to the content of the slate and thus can only be determined by the slates creator. The trigger on the other hand is a question of the context in which a slate is used and thus needs to be determined there. Does that help?

alex40724 replied:
Yes. I am not sure whether this could be reflected by a common interface accepted by the (bar) components, something that both, buttons and slates have to implement.

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *   composition: >
+	 *     The Metabar is rendered horizontally at the very top of the page. It
+	 *     is always visible and available (except in some specialized view modes
+	 *     like an exam mode) as a static screen element and is unaffected by
```
I would use "kiosk" mode instead of "exam" mode here.

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *     scrolling.
+	 *
+	 *     The Metabar always features the logo on the left hand side, while
+	 *     further elements are placed on the right hand side. Currently, these
```
It could be more flexible to have the logo not being a part of the metabar and add it to the page layout directly imo.

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *     marked with counters for new/existing notifications.
+	 *
+	 *   effect: >
+	 *     Entries in the Metabar may open a Slate when clicked. They will be set
```
This is where the confusion around metabar entries vs. bulky buttons vs. prompts and slates starts for me.

The current interface allows to add bulky buttons or prompts to the metabar. However "Entries in the Metabar may open a Slate when clicked". This does no make sense for Prompts being added to the metabar. They do not open slates, they are slates. There is currently no example for using a Prompt in the metabar, maybe this will make this clear. But for me it remains unclear if the triggerer (bulky buttons) are part of the slates or not.

klees replied:
As far as we understand the requirements for the metadata, there currently are two types of entries: one that just is a button that does something (aka "?", which adds an according slate to the tools-section of the mainbar) and buttons that open a slate. Since we decided that the label and icon/glyph should be attached to the slate, what a consumer of the interface in fact needs to add are buttons (for the first case) and slates, where in the second case the metabar internally uses the glyph/icon and the label from the slate to create an according button and wire it to a popover that contains the slate. Does this make sense?

alex40724 replied:
Yes, thanks. See also this comment: https://github.com/ILIAS-eLearning/ILIAS/pull/1335#discussion_r238260684

alex40724 commented on src/UI/Component/MainControls/Slate/Prompt.php:
```diff
@@ -0,0 +1,15 @@
+ */
+interface Prompt extends Slate
+{
+	public function	withUpdatedStatusCounter(int $count): Prompt;
```
Slates are triggered by bulky buttons, prompts are not, why is this?
Since I can pass counters to glyphs and glyphs to bulky buttons and use these as triggeres I would have expected a very different interface here. But I may miss the point with the Prompts completely.

klees replied:
We decided to introduce a Slate and a specialized slate (the Prompt) that's supposed to be those thingys that somehow "alarm" users visually with the counters. We decided to make this counter an issue of the prompt (and not the glyph) because we expect that we will need one type of prompt ("Notifications") that bundles other prompts ("Mail", "Background Task", ...) and want to add counters on this level. So the prompt uses the counter on the glyph to visualize its own counter, but owns the counter conceptually.

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *     help, or depending on requirements of the content, e.g. a local navigation.
+	 *
+	 *   composition: >
+	 *     The Mainbar holds Slates and Bulky Buttons.
```
In the metabar the entries (bulky buttons) open slates. Why not the same in the mainbar?

klees replied:
This is very similar to the metabar. In fact we have anticipated the requirement that some Bulky Button in the Mainbar just opens some page (instead of a slate) here.

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+     *
+	 *   effect: >
+     *     Clicking an entry will carry out its configured action. For slates, this
+	 *     is expanding the slate, while for links an according page opens.
```
An entry "is" a slate or an entry "opens" a slate?
Now "Links" means "Bulky Buttons"?

klees replied:
@nhaagen Lets reword here, link is a term we already use in the KS. Just do it or ask me for ideas. Thx!

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *     slate is opened above (like in overlay, not "on top of") the content.
+	 *
+	 *     The slates height equals that of the Mainbar. Also, their position will
+	 *     remain static when the page is scrolled. A button to close a slate is
```
This maybe misread as "position static" in the CSS sense, but I think you mean "fixed" here.

alex40724 commented on src/UI/Component/Glyph/Factory.php:
```diff
@@ -834,4 +834,55 @@ public function reset($action = null);
+	 *   effect: >
+	 *       Clicking this glyph will open a search dialog.
+	 *       Since the context for the Search Glyph primarily is the Metabar,
+	 *       the according search dialog will be opened as Tool in the Sidebar.
```
Sidebar means Mainbar?

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *     Tab Bar: >
+	 *       The Mainbar (and its components) shall not be used to substitute
+	 *       functionality available at objects, such as settings, members or
+	 *       learning progress. Those remain in the Tab Bar.
```
So many bars ;-) This one is not existing in the KS yet. Personally I would prefer to not add rules to non existing KS components (but I see that there are more similar entries following).

Amstutz replied:
I see the point of @alex40724 especially the "Content Actions" and "Personal Desktop" points should just move to purpose. I am not even sure if things with such a scope will ever be KS Items. The Tab Bar is less problematic and IMO might stay here for the moment. However, we have to make ourselves a mental note to rename/adapt them as soon as we have the Tab Bar.

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *        Operating elements in the bar MUST either lead to further navigational
+	 *        options within the bar (open a slate) OR actually invoke navigation, i.e.
+	 *        change the location/content of the current page.
+	 *     2: Elements in the bar MUST NOT open a modal or window.
```
"Window" is also not an existing concept in the KS. There is a "Open In New Viewport" concept in the Links component. So we may use the term Viewport here as well (if this is what is meant here).

alex40724 commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,271 @@
+	 *     3: The bar SHOULD NOT contain more than five Entries.
+	 *     4: >
+	 *       Entries in the Mainbar MUST NOT be enhanced with counters or
+	 *       other notifications drawing the user's attention.
```
"Tools must be Slates" is also a rule, right? Or does this appear somewhere else?

klees replied:
Yes, tools will need to be implemented as slates to fit here. Tools is not an KS-entry in his own right, though, and currently we do not expect it to become one. Instead, a Slate is a tool when it is put in the according section of the Mainbar via `withToolEntry`.

alex40724 commented:
Thanks a lot for this. I stopped reading through the PR since I experience some general confusion :-) This might be just caused by the general complexity of the topic.

nhaagen committed:
some typos and wordings

nhaagen committed:
moved the logo from metabar to page

nhaagen committed:
rules for slate-contents and usage

nhaagen commented:
Thank you for the comments so far.
I extendend the rules for slates, especially in the mainbar, to be more clear on what should and what should not be contained in those.

Amstutz commented on src/UI/Component/Layout/Page/Factory.php:
```diff
@@ -0,0 +1,49 @@
+	 *     2: The page MUST be rendered with a Metabar.
+	 *     3: The page MUST be rendered with a Mainbar.
+	 *     4: The page SHOULD be rendered with Breadcrumbs.
+	 *     5: The page SHOULD be rendered with a Logo.
```
In all Rules, replace Page with Standard Page

Amstutz commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,285 @@
+	 *
+	 *     Since the controls necessary for theses purposes might be quite complex,
+	 *     they are summed up in an easy to grasp Icon or Glyph in conjunction with
+	 *     a short text. Theses reductions form the entries for the mainbar, which
```
Capitalize "mainbar"

Amstutz commented on src/UI/Component/MainControls/Factory.php:
```diff
@@ -0,0 +1,285 @@
+	 *   effect: >
+	 *     Clicking an entry will carry out its configured action. For slates, this
+	 *     is expanding the slate, while for Bulky Buttons this might be just
+	 *     changing the page, e.g.
```
e.g. -> I guess there should follow some example.

Amstutz commented on src/UI/Component/MainControls/Mainbar.php:
```diff
@@ -0,0 +1,71 @@
+	 * @param Button\Bulky|Slate $entry
+	 * @throws InvalidArgumentException 	if $id is already taken
+	 */
+	public function withEntry(string $id, $entry): Mainbar;
```
Is it append or set? I guess append, so use **add**Entry.

klees replied:
I think we had a similar discussion once and decided to use `withAdditional` in that case, because we expected people to think that just `add` would mutate the object.

Amstutz replied:
Then change to withAdditionalEntry.

Amstutz commented on src/UI/Component/MainControls/Mainbar.php:
```diff
@@ -0,0 +1,71 @@
+	 * @param Button\Bulky|Slate $entry
+	 * @throws InvalidArgumentException 	if $id is already taken
+	 */
+	public function withToolEntry(string $id, $entry): Mainbar;
```
Is it append or set? I guess append, so use **add**ToolEntry

Amstutz replied:
Some as above, withAdditionalToolEntry.

Amstutz reviewed (changes requested):
Hi @nhaagen

This is huge, in size and in importance. Thanks a lot for all your effort you put into this. 

Note 1: There many places where the descriptions and work really shines. To keep my review as easy to read as possible I only pick on the points I believe can be optimized further.

Note 2 I can see why you put several individual compoents in this particular case. In enhances the readability. However, note that the main draw back will be, that we have to agree on every single component, before we can accept this. Here this probably makes sense. In other cases, this could be an issue.

Note 3: I completely ignored the implementation section for the moment. Since we will discuss only discuss the interface in the JF, I only commented that part.

I have the following questions I wish to be discussed/answered:

- ~~Why did you change from Header Bar to Metabar? I know you aim for a name better describing the functionality, but a) this not only about Meta-(Information) and b) is somewhat hard to understand. I know Header Bar has the draw-back that it does not say much about it’s function but it is easily understood. Also, it kind of follows your description of Layout Components: "Since Layout components carry - due to their nature - certain structural * decisions, they are also about the "where" of elements as opposed to * the exclusive "what" in many other components.“ I am divided an curious what others think. I do not have a strong preference. If the other do not have one either, I am fine with sticking with the proposal, since it fits well with „Mainbar“.~~ [Edit: This Question is obsolete, has been answered during @klees durin the JF discussion]


Please add/change the following points:

- [ ] All inline comments
- [ ] Page: The page MUST be rendered with content.  <— I do not understand this rule. Please clarify. What is content? Where is it? Maybe you can leave the rule as is but place some more description on this in composition.
- [ ] Page: You are missing the doc comments on all methods of the Standard Page’s interface, please add (see: #8).
- [ ] Metabar: Please add a rule that strengthen the main purpose described above somewhat. Something among the lines: „Since mainly items that pitch the user are placed in the Metabar, you SHOULD only propose items that have the nature of informing the user in this section."
- [ ] Add descriptions of all the interface functions of the Mainbar
- [ ] Currently your **with**Entry and **with**ToolEntry methods are more about appending, so use [edit, see inline disc.] **withAdditional**. However, I still think **with** could be very useful (with a (re)set like nature). So I would welcome those as well.
- [ ] Add descriptions of all the interface functions of the Metabar
- [ ] Add descriptions of all the interface functions of the Slate

@nhaagen: Again and most importantly, thanks a lot for this big contribution. It is an important milestone towards ILIAS 6.0 and I am looking forward to the JF discussion.

nhaagen committed:
updated docs and renamed withEntry to withAdditionalEntry

nhaagen committed:
removed some inline comments

nhaagen commented:
Again, thanks a lot for your comments. I tried to take care of them in the rules and interfaces.

The name "Metabar" origins from a discussion earlier this year and is on special wish from the JF-Board ;)

I know this is a big one and quite hefty in terms of impact, but I think that all of these components are easier to understand when presented together. If there are objections to certain parts, I'll gladly knock this one down to smaller PRs, but for now: you get the picture...
As to the same extend: I did not want to include an example of the Notification-Prompt in here, since that would also mean describing the Notification-Component. This, for sure, deserves a PR on its own.

matthiaskunkel commented:
Jour Fixe, 03 DEC 2018 : Thank you very much for this huge pull request and suggestion. It is a very important step towards 6.0. We highly appreciate this pull request and accept it for trunk under the following conditions:
- We accept that the page editor is allowed to use the slate.
- Please consider Timon's points from the review above.
- The description of the slate should not only allow "one word" but a very short label (e.g. 'Local Navigation')
- We should use the labels "Meta Bar" and "Main Bar".

nhaagen committed:
renamed Metabar to MetaBar, Mainbar to MainBar

nhaagen committed:
added (few) tests and styling

nhaagen committed:
clear/update composer files

nhaagen committed:
update on styles: design bulky-buttons in mainbar

nhaagen commented:
Hi,
I addressed the points from Jour Fixe, 03 DEC 2018; with the latest commit, also the looks is quite close to the sketches. There are more features for the mainbar which I have almost ready locally, but I certainly would like to decouple those from this big PR. 
Maybe you could have a look at the changes here and merge them, if there are no objections? 
I'll then continue with smaller PRs, promise ;)

Amstutz commented:
Hi @nhaagen,

Not sure, if I understand correctly. I do not like the idea of merging work in progress. However, if you give me something like a list of issues or things you know to be missing, I can perform a review on the rest and merge as soon as everything is fine there.

Greetings

nhaagen commented:
Hi @Amstutz ,
I totally agree. However, this PR is - with the exception of necessary cleanup in CSS - complete in itself.
What I would like to add is, e.g, RPC-capability for the slate to replace its contents and a  "backlink" as pictured in https://sketch.cloud/s/yLndE/Zr7Pa7, both as preliminary work for a Drilldown-Menu.
I can add this here, too, but I feel it would be easier to discuss those things seperatedly.

nhaagen committed:
cleanup and fixes for IE

nhaagen committed:
dynamic more-button and fillup for tools

klees commented on src/UI/Component/Layout/Page/Factory.php:
```diff
@@ -0,0 +1,51 @@
+	 *     2: The Standard Page MUST be rendered with a Metabar.
+	 *     3: The Standard Page MUST be rendered with a Mainbar.
+	 *     4: The Standard Page SHOULD be rendered with Breadcrumbs.
+	 *     5: The Standard Page SHOULD be rendered with a Logo.
```
The indentation in this ks-entry seems to be different in every paragraph. This hurts the little symetricist in me... Please rectify =)

klees commented on src/UI/Component/Layout/Page/Factory.php:
```diff
@@ -0,0 +1,51 @@
+		MainControls\Metabar $metabar,
+		MainControls\Mainbar $mainbar,
+		$content,
+		$locator = null,
```
Please add typehints and/or docstrings for these two arguments.

klees commented on src/UI/Component/Layout/Page/Factory.php:
```diff
@@ -0,0 +1,51 @@
+	 *
+	 * ----
+	 *
+	 * @return  \ILIAS\UI\Component\Layout\Page\Standard
```
Please just remove this. The typehint already gives this info.

klees commented on src/UI/Component/MainControls/MetaBar.php:
```diff
@@ -0,0 +1,33 @@
+	public function getEntries(): array;
+
+	/**
+	 * The Signal is triggered when any Enntry is being clicked.
```
```suggestion
	 * The Signal is triggered when any Entry is being clicked.
```

klees commented on src/UI/Component/MainControls/Slate/Factory.php:
```diff
@@ -0,0 +1,66 @@
+	 * @param string $contents
+	 * @return \ILIAS\UI\Component\MainControls\Slate\Legacy
+	 */
+	public function legacy(string $name, $symbol, string $contents): Legacy;
```
Docstrings can be removed here (once the guidelines allow this...).

klees commented on src/UI/Component/MainControls/Slate/Factory.php:
```diff
@@ -0,0 +1,66 @@
+	 * @param \ILIAS\UI\Component\Icon\Icon | \ILIAS\UI\Component\Glyph\Glyph $symbol
+	 * @return \ILIAS\UI\Component\MainControls\Slate\Combined
+	 */
+	public function combined(string $name, $symbol): Combined;
```
Docstrings can be removed here (once the guidelines allow this...).

klees commented on src/UI/Factory.php:
```diff
@@ -760,4 +760,47 @@ public function table();
+	 *
+	 * @return \ILIAS\UI\Component\Layout\Factory
+	 */
+	public function layout();
```
Add typehints (and remove docstring when allowed).

klees commented on src/UI/Factory.php:
```diff
@@ -760,4 +760,47 @@ public function table();
+	 *
+	 * @return \ILIAS\UI\Component\MainControls\Factory
+	 */
+	public function mainControls();
```
Add typehints (and remove docstring when allowed).

klees commented on src/UI/Component/Glyph/Factory.php:
```diff
@@ -834,4 +834,55 @@ public function reset($action = null);
+	 * @param	string|null	$action
+	 * @return	\ILIAS\UI\Component\Glyph\Glyph
+	 */
+	public function search($action = null);
```
Add typehints (and remove docstring when allowed).

klees commented on src/UI/Component/Glyph/Factory.php:
```diff
@@ -834,4 +834,55 @@ public function reset($action = null);
+	 * @param	string|null	$action
+	 * @return	\ILIAS\UI\Component\Glyph\Glyph
+	 */
+	public function help($action = null);
```
Add typehints (and remove docstring when allowed).

klees commented on src/UI/Implementation/Component/Layout/Page/Renderer.php:
```diff
@@ -0,0 +1,125 @@
+
+		if ($component instanceof Component\Layout\Page\Standard) {
+			return $this->renderStandardPage($component, $default_renderer);
+		}
```
Please throw a `\LogicException` if we receive an unknown component here as a safeguard and reminder.

nhaagen replied:
I thought this was done by $this->checkComponent($component) ?

klees commented on src/UI/Implementation/Component/Layout/Page/Renderer.php:
```diff
@@ -0,0 +1,125 @@
+		$tpl->setVariable('METABAR', $default_renderer->render($component->getMetabar()));
+		$tpl->setVariable('MAINBAR', $default_renderer->render($component->getMainbar()));
+
+		if($breadcrumbs = $component->getBreadcrumbs()) {
```
I think this will soon be outlawed, since this is easy to confuse with `$breadcrumbs == $component->getBreadcrumbs()`. Please move the assignment in front of the `if`.

klees commented on src/UI/Implementation/Component/Layout/Page/Renderer.php:
```diff
@@ -0,0 +1,125 @@
+			$tpl->setVariable('BREADCRUMBS', $default_renderer->render($breadcrumbs));
+		}
+
+		if($logo = $component->getLogo()) {
```
Assignment in `if` here, too.

klees commented on src/UI/Implementation/Component/Layout/Page/Renderer.php:
```diff
@@ -0,0 +1,125 @@
+	/**
+	 * When rendering the whole page, all resources must be included.
+	 * This is for now and the page-demo to work, lateron this must be replaced
+	 * with resources set as properties at the page or similar mechanisms.
```
Please transform this into an entry for the ROADMAP for better visibility.

klees commented on src/UI/Implementation/Component/Layout/Page/Renderer.php:
```diff
@@ -0,0 +1,125 @@
+	 */
+	public function registerResources(\ILIAS\UI\Implementation\Render\ResourceRegistry $registry) {
+		parent::registerResources($registry);
+	}
```
Please remove this, as it only defers to the parent, which would happen anyway...

klees commented on src/UI/Implementation/Component/MainControls/Slate/Combined.php:
```diff
@@ -0,0 +1,53 @@
+	{
+		return $this->contents;
+	}
+
```
Remove empty line, please.

klees commented on src/UI/Implementation/Component/MainControls/Slate/Combined.php:
```diff
@@ -0,0 +1,53 @@
+	 */
+	protected $contents = [];
+
+
```
Remove one empty line, please.

klees commented on src/UI/Implementation/Component/MainControls/Slate/Combined.php:
```diff
@@ -0,0 +1,53 @@
+	 */
+	public function withAdditionalEntry($entry): ISlate\Combined
+	{
+
```
Remove empty line, please.

klees commented on src/UI/templates/js/MainControls/mainbar.js:
```diff
@@ -0,0 +1,301 @@
+		var onClickToolRemoval = function(event, signalData) {
+			var search = '#' + id + ' .' + _cls_toolentries_wrapper + ' .btn',
+				active_tool_btn = $(search).filter(' .' + _cls_btn_engaged),
+				remaining;
```
The indentation screws with the readability here. I needed some time to see that this is actually a variable definition instead of a usage of `active_tool_btn`.

klees commented on tests/UI/Component/MainControls/MainBarTest.php:
```diff
@@ -0,0 +1,116 @@
+		}
+		catch (\InvalidArgumentException $e) {
+			$this->assertTrue(true);
+		}
```
Please use `this->expectException` instead. I once liked this syntax over the `@expectedException`, but the method convinced me to be better then the explicity catching.

klees reviewed (changes requested):
Hi @nhaagen (and @catkrahl implicitely)!

Thank you a lot for working so hard on the new page layout.

The content of the PR is still impressive and it took me the complete morning
to review this. Overall it looks very solid to me, I especially liked the
javascript, which for me is very readable. Even though it is readable, I still
see that we will need some conventions and shared ways for dealing with js,
which becomes obvious here once again.

In some locations your public interface does not contain docstrings at all. I
totally get why this is the case and our CaT-guidelines allow this style. The
ILIAS-guidelines, on the other hand, currently [require doc-strings](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/README.md).
I will create an according PR and argue the case on the JF. Typehints and proper
method names ftw. There are other locations that I marked where the docstrings
can vanish then as well.

Please answer the following questions:
- [x] There are multiple signals (`getEntryClickSignal`, `getToolsClickSignal`,
`getToolsRemovalSignal`) that seem to require an information on *which* entry/tool
was clicked/removed. Do I understand that correctly? How is this info passed? Where?
- [x] What is the rational behind the "_"-prefix in the js?

Please consider the following suggestions. You do not need to follow those, but
please indicate shortly why you prefer to do otherwise:
- [x] The `Slate`-interface currently has the `getContents`-method. I wonder if
this method will be implementable for every slate or if it will have a sensible
interpretation for every slate, respectively. Think about a "Chat Slate" or "Search
Slate". What would be the contents of these? I thus would suggest to remove that
method on `Slate` and instead add it with a more specific return-type on the
different, more concrete slates, if required. Since this is an interface change
we will need to go to the JF again, which would be on me. As an alternative we
might just note that issue on the ROADMAP.
- [x] Ping @chfsx about `Layout/Page/Renderer::setHeaderVars` and maybe discuss
this issue with him (if you didn't do this already). This seems to be some interesting
point for the integration.
- [x] Split `MainControls/Renderer::renderMainBar` in some more methods to improve
readability. I see `renderTools` in there, at least.
- [x] Split `MainControls\Slate\Renderer::render` in some more methods to improve
readability. I see `renderCombinedSlate` hiding somewhere in there, at least.

Please implement the following changes:
- [x] Remove the usage of the `$DIC` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1335/files#diff-6a03ba62669dff7b1bd5146e3448e5beR34).
We want to have full DI in the UI-framework. I would tolerate the usage of `$DIC`
in `Layout/Page/Renderer::setHeaderVars` because I see this going away very soon.
Not so much in the `Legacy Slate`, because there are better options here: either
inject that factory or take a `Legacy`-Component as argument. The second one would
be a slight interface change, which I would be happy to present on the JF. The same
problem is in the prompt.
- [x] Use DI for the `SignalGenerator` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1335/files#diff-f9c5461bddd130c3d60deb1836292034R356).
- [x] Add tests for behaviour of `MainBar` that is defined in the comments (`@throws
InvalidArgumentException if $id is already taken`, `@throws LogicException if no
tool-button was set`, `@throws InvalidArgumentException if $active is not an
element-identifier in entries`) and the signals. Same for `MetaBar`.

I'm really looking forward to these this in ILIAS!

kindly,
@klees

nhaagen commented:
Hi @klees,
thanks a bunch for the review. Let me answer the questions first, I'll go over the changes this afternoon.

- [x] There  are multiple signals (getEntryClickSignal, getToolsClickSignal,
getToolsRemovalSignal) that seem to require an information on which entry/tool
was clicked/removed. Do I understand that correctly? How is this info passed? Where?
- Bound functions are called with both event and signalData as params. The clicked (triggering) element is passed in signalData.triggerer, which for onClickEntry/onClickToolsEntry is the Button within the bar. 
The tools-removal-button actually searches the tools-area of the bar for anay entries.


- [x]  What is the rational behind the "_"-prefix in the js?
- Yes, that's sort of ugly. It's probably my pyhton-heritage and was meant to distinguish internal functions and private vars from others. For the functions I still like it, the css-classes-vars are pushing it a bit - but that grew and grew.... Will gladly remove, though.
 
- [x] The Slate-interface currently has the getContents-method. I wonder if
this method will be implementable for every slate or if it will have a sensible
interpretation for every slate, respectively.
- Slates are used in a somwhat complex surrounding; they have to be integrated into the bars by signals, "fix" css-classes and the like. I wanted to remove the actual contents as far as possible, so there is a single template for slates that wraps _any_ contents (as long as they are UI-Components). 
Slate-derivates can thus build up their contents any way the like and simply return them via this method.
I do not think that the (technical) slate should be an authority over the contents - and specific slates should not tolerate unwanted components in the first place.

For the rest: will do.

klees committed:
Update src/UI/Component/MainControls/MetaBar.php

Co-Authored-By: nhaagen <nhaagen@cat06.de>

nhaagen committed:
some fixes according to review

nhaagen committed:
inject factory-dependencies

nhaagen committed:
add typehints to glyphs; fix tests that are overwriting ILIAS_UI_TestBase::getUIFactory

Amstutz commented on src/UI/Implementation/Component/Layout/Factory.php:
```diff
@@ -0,0 +1,15 @@
+
+class Factory implements Layout\Factory
+{
+	public function page(): Layout\Page\Factory
```
Missing inhertitdoc

Amstutz reviewed (changes requested):
Hi @nhaagen

Thank you a lot for this major contribution to ILIAS. I think it is save to say, that this is truly a mile-stone in the history of ILIAS.

Your work is impressive. Alone doing this review took me several hours. I’ll not ask how much time it took implementing it ;-). Several of the questions or issues that popped-up are similar to the ones @klees already pointed out. I try not to duplicate.

Before we jump into the detail, I think there are two major points we need to wrap our heads around:
- The global nature of Standard Page vs. the current implementation: As rklees pointed out, it will be probably a good idea to seek contact to @chfsx (Fabian Schmid) and get his feedback (probably after the merge). In my head, I see many conflics with ilTemplate arise, but maybe I am missing several things.
- The public interface of the JS code: I am note sure if the public interface will suffice the needs that will arise (see points bellow). Note that I agree with @klees that you did a very good job in writting readable JS code, still I believe it will need some brushing as soon as we go further into the use cases (after the merge).

Now lets jump into the details ;-)

Please answer the following questions:
- [x] MessageBox Renderer Change: What has the change in src/UI/Implementation/Component/MessageBox/Renderer.php todo with any of this? If it is purely refactoring, I can live with it, but next time, better make an own small PR.
- [x] Outcommented Code in outline-glyphs: What is the story behind the outcommented code in src/UI/templates/default/Layout/outline-glyphs.less? Probably clearly document or remove.
- [x] The public interface of mainbar.js contains only 3 functions (registerSignals: registerSignals, initActive: initActive, initMore: initMore). Are you sure that is enough? Somethin like _engageButton, but a little bit extended, (e.g. like in the else clause of onClickEntry) could be really useful. E.g. When metabar stuff need to trigger mainbar stuff (e.g. help).

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] Roadmap for setHeaderVars: As klees already suggested, transform this into an entry for the ROADMAP for better visibility. Also elaborate in the entry a little, why this is important and way the page has a very special role with the header vars. Do you suggest, that all css and js has to be set by the page? Or is in fine if one still uses iltemplate him or herself for the transition period? Related to this, can you outline a little, wheter the step away from the global template to the standardpage can be made? I guess at some point you would pass the global tpl into standardpage to remain backwards compatible with class passing on load code and such? 
- [x] Default Icons, Some of the Icons used for the examples are just stored in the examples folder (e.g. Organisation or similar). What is the proposed endgame here, do you believe this to be in the responsibility of who ever implements this slate (Organisation or similar) and puts them into the mainbar to add a corresponding default icon to the core? Would be fine with me, but then we should not forget to remove the example icons here as soon as they are available in core.
- [x] Outcommented Code in outline-glyphs: What is the story behind the outcommented code in src/UI/templates/default/Layout/outline-glyphs.less? Probably clearly document or remove.
- [x] The tests seem under-developed: Especially tests for the  rendering a mostly absent. Is there a specific reason for that? Am I missing something?



Please implement the following changes:
- [x] Remove unrelated change: The PR contains an unrelated change to some phpexcel file, remove it.
- [x] Use inheritdoc: Decorate implementation functions described in the interface with inheritdocs (see inline comment)
- [ ] Visibility of example functions: Follow the suggestion of @alex40724 and move the example code the corresponding files to increase the readability.
- [x] Wrongly placed less variables: Please move all less variables from src/UI/templates/default/Layout/standardpage.less to the global variables.less files (see: https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-3/templates/Readme.md#variables). Are the colors really not extensions of existing ILIAS-Colors? Maybe check back with Christoph.
- [x] random z-index: In the same file there is a header z-index: 998, this seems quite randomly chosen (little lower than 1000), if you do use the existing bootstrap variable: @zindex-navbar and make it a bit smaller (e.g. -2). So this value gets some more semantics or introduce your own and explain in by commenting in the global variables file. Same Problem with mainbar, z-index: 995.
- [x] variable for padding of btn-bulky: In slate.less is a comment asking for a variable for this padding and some other stuff like margin and color, I would do the same :-)
- [x] variable for min-width .il-metabar-slates: The value 250px in metabar.less seems quite critical, please add a document a variable for this.
- [x] Remove Todo: remove comments: Add probably the targeted code in the correpsonding js file

As said, my two major concerns maintained in the beginning can be tackled after the merge. @chfsx: I expect, that your implementation will make some further changes on this visible. Do not hesitate to contact us as soon as you feel ready to discuss them. 

kindly,
@amstutz

nhaagen committed:
add roadmap-entry, increase legibility of slate-renderer

nhaagen committed:
weed out examples and move to proper place

nhaagen committed:
retrieve files from global template (after rebase)

nhaagen committed:
mobile layout, starting out

nhaagen committed:
fix tests for new top-level entries of ui-factory

nhaagen committed:
measure horizontal offset as well

nhaagen committed:
Never, ever, forget to update ui-factory in here, too. As a result of an error caused here, il_components is truncated during structure-reload and the instance is broken

nhaagen commented:
Hi @Amstutz,
thank you for digging into this. 

- [ ] MessageBox Renderer Change: What has the change in src/UI/Implementation/Component/MessageBox/Renderer.php todo with any of this? If it is purely refactoring, I can live with it, but next time, better make an own small PR.
- I stumbled over a lot of tests failing when introducing a top-node for the ui-factory.
This remains an issue, imo, but I tried to fix the tests at least. In doing so, the special case in MessageBox/Renderer.php became obsolete, so I ditched it.
You are right, however, this should have been a seperate PR. As there are Tree and Drilldown facing the same issue, I'll probably do it anyway.

- [ ] The public interface of mainbar.js contains only 3 functions (registerSignals: registerSignals, initActive: initActive, initMore: initMore). Are you sure that is enough? Somethin like _engageButton, but a little bit extended, (e.g. like in the else clause of onClickEntry) could be really useful. E.g. When metabar stuff need to trigger mainbar stuff (e.g. help).
- There are certainly things missing, once this is put to real action, I agree. However, at the time beeing, I find it hard to anticipate; if (when) need arises, I'd like to discuss the actual requirements and build them in. For example, I see a need to load slate-contents via AJAX...
Can we postpone this?

- [ ] Roadmap for setHeaderVars: As klees already suggested, transform this into an entry for the ROADMAP for better visibility. Also elaborate in the entry a little, why this is important and way the page has a very special role with the header vars. Do you suggest, that all css and js has to be set by the page? Or is in fine if one still uses iltemplate him or herself for the transition period? Related to this, can you outline a little, wheter the step away from the global template to the standardpage can be made? I guess at some point you would pass the global tpl into standardpage to remain backwards compatible with class passing on load code and such?
- I did the roadmap-entry. With @chfsx working on the global screen service, I guess that the oulines of CSS/JS registries will become clearer. As of now, I'd use ilGlobalTemplate and grep the files from there, but, yes, ultimately the page should be configured with the needed resources (these functions are missing in the interfaces, though).
Finally, there would be exactly one line like "echo renderer->render($page); exit();" - I would like that ;)

- [ ] Default Icons, Some of the Icons used for the examples are just stored in the examples folder (e.g. Organisation or similar). What is the proposed endgame here, do you believe this to be in the responsibility of who ever implements this slate (Organisation or similar) and puts them into the mainbar to add a corresponding default icon to the core? Would be fine with me, but then we should not forget to remove the example icons here as soon as they are available in core.
- Yes, these are meant as examples. Allthough the icons point strongly into a certain direction, I do not want to suggest a structure of the MainMenu or according visuals here.
When used, the icons should of course be pulled from delos or skins.

- [ ] The tests seem under-developed: Especially tests for the rendering a mostly absent. Is there a specific reason for that? Am I missing something?
- Nope, just my fault. I'll fix that.

- [ ] Outcommented Code in outline-glyphs: What is the story behind the outcommented code in src/UI/templates/default/Layout/outline-glyphs.less? Probably clearly document or remove.
- We first used the outline-icons on the mainbar; then, we fell back to example-images, feeling that the outline-icons would be a sepearate topic alltogether. I'll remove them for now.

- [ ]  ...CSS-related changes
- @catkrahl and me are working in parallel on this. I'll cleanup for now, and @catkrahl will either OK my changes or alter them.


I'll adress the changes and drop a note when done.
Best regards,
Nils

nhaagen committed:
remove outline-glyphs and change classname for engaged-slates at page

nhaagen committed:
renamed and moved less-variables

nhaagen committed:
z-indexes based on zindex-navbar

nhaagen commented:
Hi @catkrahl, 
I moved (and renamed) the less-variables; please have a look at the latest commits and leave a note if this is OK with you. 

https://github.com/ILIAS-eLearning/ILIAS/pull/1335/commits/16dcb693593308f29f13ab92942756ca56a7a3cb
https://github.com/ILIAS-eLearning/ILIAS/pull/1335/commits/77d9d857609ba0c9079fcaea5ce53ef19031ad74
https://github.com/ILIAS-eLearning/ILIAS/pull/1335/commits/af52051f7450cd0c28f16c3c32ecbc536bbaf512

Thanks.

catkrahl commented:
thx @nhaagen, 

the renaming and repositioning of the variables and classes was an important thing i had in mind for this PR as well! 

we probably will need a color-scheme for the slate-bulkies on the different "hierachy-levels" (e.g. for the ILIAS tree inside the slate). i used "hard-coded" colors, but there should be system with is related to the main-color-variables, such as: @brand-primary, @brand-secondary ... may we can use the color-pattern, which @Amstutz implemented. i will check that out.

shouldn't we move the z-index-variables all from the standardpage.less to the variables.less-file as well?

that was all by now. thanks and cheers
Christoph

nhaagen committed:
add basic rendering tests for maincontrols

nhaagen committed:
add legacy to dummy-factory

nhaagen commented:
From my point of view, this is good to go.
If there are objections, please write a small note and I'll try to fix immediately.
Thanks a lot.

nhaagen committed:
minor tests for the page

nhaagen committed:
throw exception when trying to set an non-existant entry to active

Amstutz reviewed (approved):
Hi Niels thx a lot for answering all the questions left. I agree, this is good to go.

Note, that two point are IMO still open:

- [x]  Remove unrelated change: The PR contains an unrelated change to some phpexcel file, remove it.
- [x]   Visibility of example functions: Follow the suggestion of @alex40724 and move the example code the corresponding files to increase the readability.

However, to me they are of such a minor importance, that I can approve this. @klees as soon as you feel that all your points have been taken care of, you may press the merge button ASAP, to avoid further conflicts ;-)

nhaagen committed:
inject counter-factory to maincontrols/slate

nhaagen committed:
added 'lost' file and added counter-mutator to raodmap

nhaagen committed:
refactored to smaller functions

nhaagen committed:
better legibility with example in one file

klees commented:
https://www.youtube.com/watch?v=1-I--8QiE84
